File: Processamento Digitais de Sinais/Aula_11/lms_teste.m
%----------------------------------------------------
% Programa para o teste do alg LMS

% Alterado : Walter Gontijo

%----------------------------------------------------
clear;  clc;close all;


% Carregando o arquivo de entrada
 fid = fopen('ruidobranco.pcm', 'rb');
 x = fread(fid, 'short');
 fclose(fid);

% Usando um valor fixo para o mu
mu = 0.000000000005;				    % Parametro de convergencia do LMS

% Comprimento dos filtros
N = 10;			            % Comprimento do filtro W(z) 
Nmax = N;		            % valor maximo dos filtros 


N_ITER = length(x);

% Vetores utilizados
wn = zeros(N,1);			% Coeficientes de W(z)
yn = zeros(Nmax,1);		    % Vetor da saida de W(z)
x_linhan = zeros(Nmax,1);	% Vetor de entrada filtrado por F^(z)


% Definindo a planta 

Wo = [.5; .3; -.5; .8];
NWo = length(Wo);
%----------------------------------------------------------


% Dados de entrada e saida
%x = load('xn.dat');		% Vetor para o sinal de entrada x(n)
%d = load('dn.dat');		% Vetor para o sinal de desejado d(n)

e_salva = zeros(N_ITER,1);		% Vetor para o sinal de erro e(n)
y_salva = zeros(N_ITER,1);		% Vetor para o sinal de controle y(n)
w_salva = zeros(N_ITER,N);      % Para plotar o comportamento dos coeficientes




  

for k=1:N_ITER
    % Lendo a amostra de entrada
    x_new = x(k,1);
   % x_new = randn;
    
    % Filtrando o sinal de entrada x
	x_linhan(1,1) = x_new;
    
    
	% Shift e atualização do vetor x_linhan
	%x_linhan = [x_linha ; x_linhan(1:size(x_linhan,1)-1,1)];
    
    

    % Calculando o vetor de y = x'*w
    %y = x_linhan(1:N,1)'*wn;
    y = 0;
    for j= 1:N
        y = y + x_linhan(j,1) * wn(j,1);
    end
    
    y_salva(k) = y;
    
   

    % Calculando a saida da planta
    %d_new = d(k);
    %d_new = x_linhan(1:length(Wo))'*Wo;  		%----> sinal desejado
    d_new = 0;
    for j= 1:NWo
        d_new = d_new + x_linhan(j,1) * Wo(j,1);
    end
    
    
    % Cálculo do erro e(n) = d(n) - y(n)
     ee = d_new - y;
  
    e_salva(k) = ee; %Salvando o erro 
    
   
    
    % Atualizando os coeficientes do filtro usando LMS
    %wn = wn + mu*ee*x_linhan;
     for j= 1:N
        wn(j,1) = wn(j,1) + mu*ee*x_linhan(j,1);
    end
    

    w_salva(k,:) = wn;  
    
    % Atualizando o vetor x_linhan
     for j= N:-1:2
        x_linhan(j,1) = x_linhan(j-1,1);
    end
    
end

% Plotando os sinais de erro e de controle
figure(1);
plot(e_salva);grid;
title('Sinal de Erro');
xlabel('Amostras');
ylabel('Erro');

figure(2)
plot(w_salva);grid;
title('Comportamento dos Coefs');
xlabel('Amostras');

File: Processamento Digitais de Sinais/Aula_11/lms_teste.c
#include <stdio.h>
//#define N 4
#define N 160

int main(){

    FILE *Data_in, *Data_out;

    if((Data_in = fopen("white_noise.pcm", "rb")) == NULL){

        printf("O arquivo de entrada nao abriu\n");
        return 0;
    }

    if((Data_out = fopen("saida_lms.pcm", "wb")) == NULL){

        printf("O arquivo de saida nao abriu\n");
        return 0;
    }

    double u = 0.000000000005; // Taxa de aprendizado
    int i, sampleN;             // Contador, Nº de amostras
    double wn[N];               // Coeficiente adaptados
    double y = 0.0;             // Saída do filtro adaptativo
    double d = 0.0;             // Coeficientes desejados
    double error;               // Erro 
    // Amostras, entrada, saida e erro em short
    short sample[N], input, output, shortError;

    for(i = 0; i < N; i++)
    {
        wn[i] = 0.0;
        sample[i] = 0.0;
    }

    // Definindo a planta

/*     float Wo[]={
   				#include "coef_pb.dat"
    }; */

    
    float Wo[]={
   				#include "coef_pa.dat"
    };
    
    /* double Wo[] = {0.5, 0.3, -0.5, 0.8}; */

    do{
        d = 0;
        y = 0;
        // Lendo a amostra de entrada
        sampleN = fread(&input, sizeof(short), 1, Data_in);
        sample[0] = input;   // Entrada

        // Convolução de y(n) da entrada com o vetor de coeficientes atualizados
        for(i = 0; i < N; i++)
        {
            y += sample[i] * wn[i];
        }

        // d(n) = convolução da entrada com o sistema desconhecido
        for(i = 0; i < N; i++)
        {
            d += sample[i] * Wo[i];
        }

        // Calculo do erro e(n) = d(n) - y(n)
        // Diferença entre saída do sistema desconhecido e o filtro adaptativo
        error = d - y;

        // Atualizando os coeficientes do filtro usando LMS
        for(i = 0; i < N; i++)
        {
            wn[i] += 2 * u * error * sample[i];   // Sinal de erro retorna ao filtro, e calcula novos coeficientes
        }
        // Atualizando o vetor x_linhan (DESLOCAMENTO)
        for(i = N; i >= 1; i--)
        {
            sample[i] = sample[i-1];
        }

        shortError = (short) error;
        //shortError = (short) y;

        fwrite(&shortError, sizeof(short), 1, Data_out);
        //printf("\t e: %f\n", ee);

    }while(sampleN);

    printf("\n---------- RESULTADO ----------\n");
    for(i = 0; i < N; i++)
    {
        printf("W[%d]: %f\n", i, wn[i]);
    }

    fclose(Data_in);
    fclose(Data_out);

    return 0;
}
